fastlane_core/lib/fastlane_core/itunes_transporter.rb

```diff
@@ -180,9 +180,123 @@ def additional_upload_parameters
     end
   end
 
+  # Generates commands and executes the altool.
+  class AltoolTransporterExecutor < TransporterExecutor
+    ERROR_REGEX = /\sNSLocalizedFailureReason\s=\s"+(.+)"/
+
+    private_constant :ERROR_REGEX
+
+    def execute(command, hide_output)
+      if Helper.test?
+        yield(nil) if block_given?
+        command
+      end
+
+      @errors = []
+      @warnings = []
+      @all_lines = []
+
+      if hide_output
+        # Show a one time message instead
+        UI.success("Waiting for App Store Connect transporter to be finished.")
+        UI.success("Application Loader progress... this might take a few minutes...")
+      end
+
+      begin
+        exit_status = FastlaneCore::FastlanePty.spawn(command) do |command_stdout, command_stdin, pid|
+          begin
+            command_stdout.each do |line|
+              @all_lines << line
+              parse_line(line, hide_output) # this is where the parsing happens
+            end
+          end
+        end
+      rescue => ex
+        # FastlanePty adds exit_status on to StandardError so every error will have a status code
+        exit_status = ex.exit_status
+        @errors << ex.to_s
+      end
+
+      @errors << "The call to the altool completed with a non-zero exit status: #{exit_status}. This indicates a failure." unless exit_status.zero?
+
+      UI.important(@warnings.join("\n")) unless @warnings.empty?
+
+      if @errors.count > 0 && @all_lines.count > 0
+        # Print out the last 18 lines, this is key for non-verbose mode
+        @all_lines.last(18).each do |line|
+          UI.important("[altool] #{line}")
+        end
+        UI.message("Application Loader output above ^")
+        UI.error(@errors.join("\n"))
+      end
+
+      yield(@all_lines) if block_given?
+      exit_status.zero?
+    end
+
+    def build_upload_command(username, password, source = "/tmp", provider_short_name = "", jwt = nil, platform = nil, api_key = nil)
+      [
+        ("API_PRIVATE_KEYS_DIR=#{api_key[:key_filepath]}" if api_key),
+        "xcrun altool",
+        "--upload-app",
+        ("-u #{username.shellescape}" unless api_key),
```

Comment:
Hi @freddi-kit 

Nice work! i am just testing this PR and it got crashed in this line `shellescape` 
Also, the username is indeed nil in my case because i am using API-Key inside pilot so technially this line should not even execute. i think, maybe we are not passing the api_key correct way in this method....

```ruby
3: from /Users/distiller/project/vendor/bundle/ruby/2.7.0/bundler/gems/fastlane-ed28c906445c/fastlane/lib/fastlane/actions/upload_to_testflight.rb:34:in `run'
	 2: from /Users/distiller/project/vendor/bundle/ruby/2.7.0/bundler/gems/fastlane-ed28c906445c/pilot/lib/pilot/build_manager.rb:50:in `upload'
	 1: from /Users/distiller/project/vendor/bundle/ruby/2.7.0/bundler/gems/fastlane-ed28c906445c/fastlane_core/lib/fastlane_core/itunes_transporter.rb:697:in `upload'
/Users/distiller/project/vendor/bundle/ruby/2.7.0/bundler/gems/fastlane-ed28c906445c/fastlane_core/lib/fastlane_core/itunes_transporter.rb:242:in `build_upload_command': undefined method `shellescape' for nil:NilClass (NoMethodError)

Reply:
btw, `pilot + username(apple-id), password` is working well 🚀

Reply:
Oh Thank you for reviewing!

in my environment, pilot+api_key works. and here is my example Fastfile.
How is mine and your Fastfile different?

```ruby
default_platform(:ios)

platform :ios do
  desc "Description of what the lane does"
  lane :custom_lane do
    xcode_select("/Applications/Xcode_14 beta 6.app")
    api_key = app_store_connect_api_key(
      key_id: "XXXXXX",
      issuer_id: "XXX-XXX-XXXX",
      key_filepath: "./AuthKey_XXXX.p8",
      duration: 1200,
      in_house: false
    )    
    
    pilot(api_key: api_key)
  end
end
```

Reply:
Ah... sorry I may found there is a issue to check password is available or not, please wait a moment

Reply:
I noticed this is simple issue comes from my less experience Ruby 🙇 I can fix it soon

Reply:
thank you very much for putting in all the hard work! 🌟

Reply:
https://github.com/fastlane/fastlane/pull/20631/commits/711eb3cd3835eb6068a7b2a4bc29680c6650b7ab fixed, could you try it?

Reply:
I just switched over to this `altool-upload` branch in my Gemfile and I am still hitting this issue. One difference I notice is that I'm specifying `api_key_path` instead of `api_key`

Reply:
Hi @freddi-kit! Thank you so much for working on that.

I second @raid5 comment. I believe there's an unhandled case when using `api_key_path` (instead of `api_key`):

> 
>
> fastlane_core/lib/fastlane_core/itunes_transporter.rb:241:in `build_upload_command': undefined method `shellescape' for nil:NilClass (NoMethodError)

Reply:
I just investigated and it is bug, and it requires a little more changes to support it api_key_path. I'm happy if you wait for it 🙇

Reply:
Fixed! https://github.com/fastlane/fastlane/pull/20631/commits/22cfcec0f28ac400272b27ccc27de7cbf7ebcc99

Reply:
Awesome this works great! I was able to successfully upload to TestFlight now.